main.c

```diff
@@ -0,0 +1,72 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include "mpi.h"
+#include "cuda.h"
+#include "cuda_runtime.h"
+
+// Julia headers (for initialization and gc commands)
+#include "uv.h"
+#include "julia.h"
+#include "diffusion.h"
+
+JULIA_DEFINE_FAST_TLS()
+
+int main(int argc, char *argv[])
+{
+  // Initialization of libuv and julia
+  uv_setup_args(argc, argv);
+  libsupport_init();
+  jl_parse_opts(&argc, &argv);
+  jl_options.image_file = JULIAC_PROGRAM_LIBNAME;  // JULIAC_PROGRAM_LIBNAME defined on command-line for compilation
+  julia_init(JL_IMAGE_JULIA_HOME);
+
+  const int NDIMS = 2;
+  int ret, nprocs, me, reorder = 0;
+  int dims[] = {0,0}, coords[] = {0,0}, periods[] = {0,0};
+  MPI_Comm comm=MPI_COMM_NULL;
+
+  // Numerics
+  size_t nx     = 64;      // Number of grid points in x dimension
+  size_t ny     = 64;      // Number of grid points in y dimension
+  size_t nt     = 1000;    // Number of time steps
+
+  // Physics
+  double lx     = 10.0;    // Length of domain in x dimension
+  double ly     = 10.0;    // Length of domain in y dimension
+  double lam    = 1.0;     // Thermal conductivity
+  double cp_min = 1.0;     // Minimal heat capacity
+  double amp_T  = 100.0;   // Amplitude of Gaussian anomly in temperature field
+
+  double *Cp = (double *)malloc(nx*ny*sizeof(double));  // Heat capacity field
+  double *Cp_d;
+
+  // Create Cartesian process topology
+  MPI_Init(&argc, &argv);
```

Comment:
It isn't an issue with MPICH or its derivatives (including Cray MPI), but for Open MPI we would need a way to trigger the `mpi_init_hooks` in MPI.jl when initializing externally:
https://github.com/JuliaParallel/MPI.jl/issues/432

Reply:
I've tagged a new version.

Reply:
Awesome, thanks a lot!

Reply:
Yes.

Reply:
You will also need to call `MPI.run_init_hooks()` to support Open MPI.